system/jlib/jtrace.hpp

```diff
@@ -18,21 +18,197 @@
 #ifndef JTRACE_HPP
 #define JTRACE_HPP
 
+#undef UNIMPLEMENTED //opentelemetry defines UNIMPLEMENTED
+#include "opentelemetry/exporters/ostream/span_exporter_factory.h"
+#include "opentelemetry/sdk/trace/exporter.h"
+#include "opentelemetry/sdk/trace/processor.h"
+#include "opentelemetry/sdk/trace/simple_processor_factory.h"
+#include "opentelemetry/sdk/trace/tracer_context.h"
+#include "opentelemetry/sdk/trace/tracer_context_factory.h"
+#include "opentelemetry/sdk/trace/tracer_provider_factory.h"
+#include "opentelemetry/trace/provider.h"
+
+#include "opentelemetry/context/propagation/global_propagator.h"
+#include "opentelemetry/context/propagation/text_map_propagator.h"
+#include "opentelemetry/trace/propagation/http_trace_context.h"
+
+#include "opentelemetry/ext/http/client/http_client_factory.h"
+#include "opentelemetry/ext/http/common/url_parser.h"
+#include "opentelemetry/trace/semantic_conventions.h"
+
+#include <opentelemetry/context/context.h>
+
+//using namespace opentelemetry::trace;
+namespace http_client = opentelemetry::ext::http::client;
+namespace context     = opentelemetry::context;
+namespace nostd       = opentelemetry::nostd;
+namespace trace_sdk   = opentelemetry::sdk::trace;
+namespace opentel_trace = opentelemetry::trace;
+
+//#include "jexcept.hpp" //re-define UNIMPLEMENTED
+#define UNIMPLEMENTED throw makeStringExceptionV(-1, "UNIMPLEMENTED feature at %s(%d)", sanitizeSourceFile(__FILE__), __LINE__)
+#include "jprop.hpp"
+
+// TextMapCarrier is the storage medium used by TextMapPropagator.
+// pure virtual Get(key) returns the value associated with the passed key.
+// pure virtual Set(key, value) stores the key-value pair.
+// virtual Keys(nostd::function_ref<bool(nostd::string_view)> /* callback */) 
+//         list of all the keys in the carrier.
+//         By default, it returns true without invoking callback */
+template <typename T>
```

Comment:
This looks wrong being a template class - it suggests a design decision hasn't been made, and requires everything is in the header - which will cause compile sloath.

Reply:
This largely as-is example from OpenTel doesn't look wrong to me given what they were hoping to accomplish (get,set,keys for various representations of "httpheaders"), but we can provide certainly go a different way (do I recall you mentioning a lambda function interface )?